network/basics/uri.py

```diff
@@ -333,7 +356,16 @@ def uri(module, url, dest, body, body_format, method, headers, socket_timeout):
         # Reset follow_redirects back to the stashed value
         module.params['follow_redirects'] = follow_redirects
 
-    resp, info = fetch_url(module, url, data=body, headers=headers,
+
+    if source:
+        try:
+            body = load_binary(source)
+            resp, info = fetch_url(module, url, data=body.read(), headers=headers,
+                           method=method, timeout=socket_timeout)
+        finally:
+            body.close()
+    else:
+        resp, info = fetch_url(module, url, data=body, headers=headers,
```

Comment:
Seeing as we have to read the entire source anyhow to send it to fetch_url() and that the load_binary() function is so trivial... maybe this section would be better like this:
``` python

if source:
    try:
        fh = open(source, 'rb')
        body = fh.read()
    except (OSError, IOError):
        module.fail_json(msg='Unable to read the source file %s' % (to_native(source)))
    finally:
        fh.close()
resp, info = fetch_url(module, url, data=body, headers=headers, method=method, timeout=socket_timeout)
```

(Note: import to_native on line 237 as well)